perception/alars/auv_detector/auv_detector/hook_estimator.py

```diff
@@ -0,0 +1,1020 @@
+            self.listener_callback,            
+            10)
+        self.subscription_camera  # prevent unused variable warning
+
+        self.bridge = CvBridge()
+
+        self.cv_image = None
+        self.stamp = None
+        self.debug_imshow = True
+
+        self.hook_timer = self.create_timer(0.1, self.process_hook_estimation)
+
+        self.subscription_accel = self.create_subscription(
+            Imu,
+            '/Quadrotor/core/imu',
+            self.imu_callback,
+            10)
+
+        # Initialization (in __init__ or once)
+        self.est_Hook_pos_in_winch_frame = np.array([np.nan, np.nan, np.nan])
+        self.est_Hook_pos_in_base_link_frame = np.array([np.nan, np.nan, np.nan])
+        self.ekf_Hook_pos_in_winch_frame = np.array([np.nan, np.nan, np.nan])
+        self.ekf_Hook_pos_in_base_link_frame = np.array([np.nan, np.nan, np.nan])
+        self.est_Hook_pos_in_winch_frame_ekf = np.array([np.nan, np.nan, np.nan])
+        self.Hook_pos_in_base_link_frame_to_plot = deque(maxlen=2000)
+        self.est_Hook_pos_in_base_link_frame_to_plot = deque(maxlen=2000)
+        self.ekf_Hook_pos_in_base_link_frame_to_plot = deque(maxlen=2000)
+        self.est_hook_timestamps = deque(maxlen=2000)  # Store timestamps of hook detections
+        self.camera_sees_hook = deque(maxlen=2000)
+
+        self.plot_initialized = False
+        self.ekf_est_hook_map_pos_to_save = deque(maxlen=2000)  # Store EKF hook estimated pos
+        self.ekf_initialized = False
+        self.last_ekf_time = None
+
+        self.current_length = 5.0 # Length of the rope in meters, can be changed by winch control
+
+        # EKF setup
+        self.ekf = ExtendedKalmanFilter(dim_x=6, dim_z=3)  # State: [p, v]; Measured: [r]
+        self.ekf.x = np.zeros(6)
+        self.ekf.P *= 1e-1
+        self.ekf.R = np.eye(3) * 1e-1                       # Measurement noise (camera)
+        self.ekf.Q = np.eye(6) * 1e-2                       # Process noise
+        self.g_vec = np.array([0, 0, 0])  # Gravity vector in base_link frame
+
+        # Initialize IMU values
+        self.a_drone = None
+        self.smoothed_accel = None
+        self.a_drone_odom = None
+
+        # Initialize the tf buffer and listener
+        self.tf_buffer = Buffer()
+        self.tf_listener = TransformListener(self.tf_buffer, self)
+
+        # Set frames
+        self.base_link_frame = 'Quadrotor/base_link'
+        self.odom_frame = 'Quadrotor/odom'
+        self.camera_frame = 'Quadrotor/camera_link'
+        self.winch_frame = 'Quadrotor/winch_link'
+        self.imu_frame = 'Quadrotor/imu_link'
+        self.hook_frame = 'Quadrotor/Hook'
+
+        self.tf_buffer = Buffer()
+        self.tf_listener = TransformListener(self.tf_buffer, self)
+
+        self.static_tf_timer = self.create_timer(2.0, self.init_static_transfroms_once)
+
+        self.est_hook_pub = self.create_publisher(PointStamped, '/hook_est_position', 10)
```

Comment:
add your new topics to dji_msgs topics. follow the naming scheme in the file. include "ESTIMATE" for things you estimate :)